Sources/MapboxCoreNavigation/RouterDelegate.swift

```diff
@@ -1,6 +1,27 @@
 import Foundation
 import CoreLocation
 import MapboxDirections
+import Turf
+
+/**
+ Configuration for offsetting first route maneuver during rerouting.
+ */
+public enum ReroutingManeuverOffset {
+    /**
+     Leaves original `RouteOptions.initialManeuverAvoidanceRadius` to be used for rerouting attempt.
+     */
+    case `default`
+    /**
+     Disables offsetting the maneuver.
+     */
+    case disabled
```

Comment:
What’s the difference between `disabled` and `radius(0)`? Could we simplify this type to `LocationDistance?`, where `nil` is the default behavior and a value of 0 disables the behavior?

Reply:
Directions will report an error if user tries to send `radius(0)`. `disabled` is a proper way to disable the feature. I decided to use an enum for this to a) make it explicit when feature is disabled or enabled and b) provide a clean explanation what measure is used for calculation. It will be much easier to introduce new cases like `distanceAhead` or `numUpcomingIntersections` or anything else if needed.

Reply:
I can see how with this enum, we want to present 3 states of overriding `RouteOptions.initialManeuverAvoidanceRadius`:
- don't override (use original value, that was set during Directions request), (1)
- override:
  - disable buffer whatsoever, (2)
  - override buffer with a concrete radius value. (3)

And it looks kind of like `Optional(Optional(LocationDistance))` type, where external `nil` would represent NavigationSDK-default behavior (reuse the original value) and internal `nil` would represent directions-swift-default behavior (don't account for any radius).

While this is a justified approach, I would find it more simple and clear, if for the state (2) we could use value `0`.

This applies to directions-swift too, as `0` is semantically an equivalent to `nil` in `RouteOptions.initialManeuverAvoidanceRadius`, and they both mean "don't take any buffer into account".
However, if we set `.initialManeuverAvoidanceRadius` to `0`, we will actually [take a buffer of radius `1` into account](https://github.com/mapbox/mapbox-directions-swift/pull/609/files#diff-b186eea493906feb8b8810201920ee0e5d06a8cf5ff1085728c52cfd70715267R209-R212), while building a route.

In short, as [suggested in Minh's comment](https://github.com/mapbox/mapbox-navigation-ios/pull/3454/files#r747157165), I think we could use `LocationDistance?` in place of this type, but for that to work, we should allow for `0` value in `RouteOptions.initialManeuverAvoidanceRadius` and make it equivalent to `nil`.

Even if we stay with this better-described type, I think the change to directions-swift will help avoid ambiguity in using this property.

Reply:
Using `LocationDistance?` instead of `ReroutingManeuverBuffer` creates ambiguity in next way: `nil` is used in `RouteOptions` to disable the feature, while `nil` for rerouting actually means "do not modify RouteOptions value". Even more - it may look like a bug if user compares 2 cases:
1. `RouteOptions.initialManeuverAvoidanceRadius = nil` + `initialManeuverBufferWhenReroutingFrom = nil` -> feature is disabled on a reroute;
2. `RouteOptions.initialManeuverAvoidanceRadius = 100` + `initialManeuverBufferWhenReroutingFrom = nil` -> radius of 100m will be calculated.

Thus, same delegate return value "seem" to have different result.

Using `0` as a value to disable the feature also does not align with current implementation of `RouteOption`'s equivalent.
Given these mismatches I see it is confusing to try to use bare optional. IMHO an enum with explicit and documented options is a good choice to clarify user's intent.

Reply:
> nil is used in RouteOptions to disable the feature

Technically, RouteOptions uses `nil` for the Directions API’s default behavior, which happens to be disabled: https://github.com/mapbox/mapbox-navigation-ios/pull/3454#discussion_r747159518. It’s akin to the difference between RouteOptions defaulting `includesSteps` to `false` but NavigationRouteOptions defaulting it to `true`, because _navigation_ route options are more nuanced. Doesn’t the “when rerouting” in `initialManeuverBufferWhenReroutingFrom` give us enough leeway to introduce different semantics?

Reply:
I've uploaded implementation variant with using bare optional. Note [delegate method comment](https://github.com/mapbox/mapbox-navigation-ios/pull/3454/files#diff-6d9228a7ef618b724af839a609c163b32a2f6e9ec90a797ce55d881ac60b9540R48) is now responsible for explaining it's return value.
I still believe that `enum` would work much better to cover that use case.